### ISSUE TITLE
SubRange of XMLIndexer

### ISSUE THREAD

**Issue by LordNali** [drmohundro/SWXMLHash#170]
Hi,

I've got a question in regard to the (usage of the) XMLIndexer. A small example to illustrate our 'problem':
     
    <FOO>
        <item type="foo1"/>
        <keyword type="BAR"/>
        <datatype type="bar1"/>
        <layoutType type="bar2"/>
        <addrType type="bar3"/>
    </FOO>

This xml-snippet should be in the translate to the following Swift Structure

    struct FooObject {
        var itemType: String = "foo1"
        var bar: BarObject
    }

    struct BarObject {
        var datatype: String = "bar1"
        var layoutType: String = "bar2"
        var addrType: Stirng = "bar3"
     }

- First thing to know: We have no way of changing the format / structure of the XML. So unfortunately we have to use it as it is.
- Name, Types and Values are non-relevant here ;-)
- For the conversion of the xml-based-objects to their swift-based-counterparts we implemented the `XMLIndexerDeserializable` for `FooObject` and `BarObject`.
- Our problem now lies in the deserialization of the `BarObject`.

- We're using the XMLElement `keyword`to detect a potential `BarObject`. Then we know: "Okay, now there's a BarObject coming next".  So we're expecting 3 more child `XMLElement` that specify the `BarObject`. So in the end we're doing this:

       static func deserialize(_ indexer: XMLIndexer) throws -> FooObject {

           // Process FooObject here
           // Works perfectly ;-)
           ...
 
           // Now look for the keyword indicating the occurrence of a BarObject
           for child in indexer.children {

               if child.attribute(by: "type")?.text == "BAR" {
                   // BarObject detected
                   let barObject: BarObject = try! indexer.value()
               }

           }

      }

So the last actual statement here `let barObject: BarObject = try! indexer.value()`will fail, because the `BarObject`deserializer gets feeded the complete XMLIndexer (which include the children before the `<keyword>`element, and anything after that wich might not be related to the `BarObject`). What we would need is a way of feeding the `BarObject XMLIndexer` a subrange of the `FooObject` `XMLIndexer` - so only the parts of the `XMLIndexer` that are actually relevant for the `BarObject`.

Long story short: Is there a way of creating a `XMLIndexer` based on an existing one, but restricting the range of the available children of this `XMLIndexer`?

To be clear, I know this would be much easier if we would have something likes this - but unfortunately, we're not able to change the structure of the XML.

    <FOO>
        <item type="foo1"/>
        <keyword type="BAR">
            <datatype type="bar1"/>
            <layoutType type="bar2"/>
            <addrType type="bar3"/>
        </keyword>
    </FOO>

**Comment by drmohundro** [comment]
First off, thanks for the detail - I think I understand exactly what you're asking :)

## Possibility 1 - Handle Deserialization in just `FooObject` for both structs

Is the relationship between `FOO` and `BAR` in your example always 1 to 1? If so, I think my suggestion would be to only have `FooObject` implement `XMLIndexerDeserializable` and then handle build the `BarObject` in `FooObject`'s `deserialize` method as you should have access to all of the elements. Now, that clearly isn't following SRP quite as well, but it seems the most straightforward given your example.

## Possibility 2 - Maybe Create an `XMLIndexer` by hand

That being said, I do realize that this is probably a simplified example, so I'll try to answer your question of "can I create an `XMLIndexer` from another `XMLIndexer`". To be honest, there isn't an easy way. The only way I can think to do it at the moment is to manually new up an `XMLIndexer` with an `XMLElement` that you create. I think all of the relevant properties and methods are public, so it should work.

Here's a list of the relevant bits of code:

* [XMLIndexer's init](https://github.com/drmohundro/SWXMLHash/blob/master/Source/SWXMLHash.swift#L596)
* [XMLElement class](https://github.com/drmohundro/SWXMLHash/blob/master/Source/SWXMLHash.swift#L774)
* [XMLElement.children](https://github.com/drmohundro/SWXMLHash/blob/master/Source/SWXMLHash.swift#L815)

Given that you'll already have a populated `XMLElement`, it might not be too big of a deal to just remove the relevant elements via the `children` property. The problem is that the core parts of SWXMLHash are classes instead of structs, so you might run into some referential issues if you modify the underlying `children` list.

Here's some untested code:

```swift
// grab the element property from the indexer
let element = indexer.element!

// get rid of the element you're not wanting
element.children.remove(at: 0)

// create a new XMLIndexer with your new element
let subIndexer = XMLIndexer(element)
```

I hope this is helpful!

**Comment by LordNali** [comment]
Thank's for the feedback. So Possibility 1 is our current "fix".

I'll try figuring out Possibility 2 in the meanwhile, the idea seems clear. I'll post my findings then.

**Comment by drmohundro** [comment]
FYI, I just pushed `4.4.0` for this feature. Thanks again for the suggestion and the work to get this set up!
